File: src/backend/docker.ts
import { Channel, invoke } from '@tauri-apps/api/core';

// ─── SHARED TYPES ─────────────────────────────────────────────────────────────
export interface OkResponse {
  ok: boolean;
  message?: string;
}

export interface PruneResult {
  deleted: string[];
  reclaimedBytes: number;
  reclaimedHuman: string;
}

export interface CommandError {
  message: string;
  code?: string;
}

// ─── SYSTEM ───────────────────────────────────────────────────────────────────
export interface SystemInfo {
  dockerVersion: string;
  apiVersion: string;
  os: string;
  arch: string;
  totalMemory: number;
  cpuCount: number;
  containersRunning: number;
  containersStopped: number;
  containersPaused: number;
  imagesCount: number;
  serverVersion: string;
  kernelVersion: string;
  operatingSystem: string;
  storageDriver: string;
}

export interface DiskUsage {
  containersSize: number;
  imagesSize: number;
  volumesSize: number;
  buildCacheSize: number;
  totalSize: number;
}

export const dockerPing = () => invoke<boolean>('docker_ping');
export const dockerReconnect = () => invoke<boolean>('docker_reconnect');
export const dockerSystemInfo = () => invoke<SystemInfo>('docker_system_info');
export const dockerDiskUsage = () => invoke<DiskUsage>('docker_disk_usage');

// ─── CONTAINERS ───────────────────────────────────────────────────────────────

export type ContainerStatus =
  | 'running'
  | 'stopped'
  | 'paused'
  | 'exited'
  | 'restarting'
  | 'dead'
  | 'created'
  | 'removing';

export type RestartPolicy = 'no' | 'always' | 'on-failure' | 'unless-stopped';

export interface PortMapping {
  hostPort: string;
  containerPort: string;
  protocol: string;
}

export interface ContainerSummary {
  id: string;
  shortId: string;
  name: string;
  image: string;
  status: ContainerStatus;
  state: string;
  ports: PortMapping[];
  cpuPercent: number;
  memoryUsage: number;
  memoryLimit: number;
  memoryHuman: string;
  uptime: string;
  created: number;
  labels: Record<string, string>;
}

export interface MountPoint {
  source: string;
  destination: string;
  mode: string;
  rw: boolean;
  mountType: string;
}

export interface NetworkEndpoint {
  ipAddress: string;
  gateway: string;
  macAddress: string;
}

export interface ContainerInspect {
  id: string;
  name: string;
  image: string;
  imageId: string;
  status: ContainerStatus;
  created: string;
  startedAt: string;
  finishedAt: string;
  restartCount: number;
  platform: string;
  environment: string[];
  cmd: string[];
  entrypoint: string[];
  workingDir: string;
  ports: PortMapping[];
  mounts: MountPoint[];
  networks: Record<string, NetworkEndpoint>;
  labels: Record<string, string>;
  hostname: string;
  ipAddress: string;
}

export interface ContainerStats {
  id: string;
  cpuPercent: number;
  memoryUsage: number;
  memoryLimit: number;
  memoryPercent: number;
  networkRx: number;
  networkTx: number;
  blockRead: number;
  blockWrite: number;
  pids: number;
}

export interface RunContainerOptions {
  image: string;
  name?: string;
  ports: PortMapping[];
  env: string[];
  cmd?: string[];
  volumes: string[];
  detach: boolean;
  autoRemove: boolean;
  restartPolicy: RestartPolicy;
  network?: string;
  labels: Record<string, string>;
}

export interface LogLine {
  containerId: string;
  containerName: string;
  stream: 'stdout' | 'stderr';
  message: string;
  timestamp: string;
}

export const listContainers = (all?: boolean) =>
  invoke<ContainerSummary[]>('list_containers', { all });
export const inspectContainer = (id: string) =>
  invoke<ContainerInspect>('inspect_container', { id });
export const startContainer = (id: string) =>
  invoke<OkResponse>('start_container', { id });
export const stopContainer = (id: string, timeout?: number) =>
  invoke<OkResponse>('stop_container', { id, timeout });
export const restartContainer = (id: string, timeout?: number) =>
  invoke<OkResponse>('restart_container', { id, timeout });
export const pauseContainer = (id: string) =>
  invoke<OkResponse>('pause_container', { id });
export const unpauseContainer = (id: string) =>
  invoke<OkResponse>('unpause_container', { id });
export const killContainer = (id: string, signal?: string) =>
  invoke<OkResponse>('kill_container', { id, signal });
export const removeContainer = (id: string, force?: boolean) =>
  invoke<OkResponse>('remove_container', { id, force });
export const renameContainer = (id: string, newName: string) =>
  invoke<OkResponse>('rename_container', { id, newName });
export const runContainer = (options: RunContainerOptions) =>
  invoke<string>('run_container', { options });
export const getContainerStats = (id: string) =>
  invoke<ContainerStats>('get_container_stats', { id });
export const getContainerLogs = (id: string, tail?: number, since?: number) =>
  invoke<LogLine[]>('get_container_logs', { id, tail, since });

// ─── IMAGES ───────────────────────────────────────────────────────────────────

export interface ImageSummary {
  id: string;
  shortId: string;
  repository: string;
  tag: string;
  digest: string;
  size: number;
  sizeHuman: string;
  created: number;
  inUse: boolean;
  architecture: string;
  os: string;
  labels: Record<string, string>;
  containers: string[];
}

export const listImages = () => invoke<ImageSummary[]>('list_images');
export const inspectImage = (id: string) =>
  invoke<ImageSummary>('inspect_image', { id });
export const pullImage = (image: string, tag?: string) =>
  invoke<OkResponse>('pull_image', { image, tag });
export const removeImage = (id: string, force?: boolean) =>
  invoke<OkResponse>('remove_image', { id, force });
export const pruneImages = () => invoke<PruneResult>('prune_images');

// ─── VOLUMES ──────────────────────────────────────────────────────────────────

export interface VolumeSummary {
  name: string;
  driver: string;
  mountpoint: string;
  scope: string;
  created: string;
  labels: Record<string, string>;
  options: Record<string, string>;
  inUse: boolean;
  containers: string[];
  size?: number;
  sizeHuman: string;
}

export const listVolumes = () => invoke<VolumeSummary[]>('list_volumes');
export const createVolume = (
  name: string,
  driver?: string,
  labels?: Record<string, string>,
  driverOpts?: Record<string, string>
) =>
  invoke<VolumeSummary>('create_volume', { name, driver, labels, driverOpts });
export const removeVolume = (name: string, force?: boolean) =>
  invoke<OkResponse>('remove_volume', { name, force });
export const pruneVolumes = () => invoke<PruneResult>('prune_volumes');

// ─── NETWORKS ─────────────────────────────────────────────────────────────────

export interface NetworkContainerEntry {
  name: string;
  ip: string;
  macAddress: string;
}

export interface NetworkSummary {
  id: string;
  shortId: string;
  name: string;
  driver: string;
  scope: string;
  subnet: string;
  gateway: string;
  ipRange: string;
  internal: boolean;
  attachable: boolean;
  created: string;
  labels: Record<string, string>;
  isDefault: boolean;
  containers: NetworkContainerEntry[];
}

export interface CreateNetworkOptions {
  name: string;
  driver: string;
  subnet?: string;
  gateway?: string;
  ipRange?: string;
  internal: boolean;
  attachable: boolean;
  labels: Record<string, string>;
}

export const listNetworks = () => invoke<NetworkSummary[]>('list_networks');
export const createNetwork = (options: CreateNetworkOptions) =>
  invoke<string>('create_network', { options });
export const removeNetwork = (id: string) =>
  invoke<OkResponse>('remove_network', { id });
export const connectContainerToNetwork = (
  networkId: string,
  containerId: string,
  ip?: string
) =>
  invoke<OkResponse>('connect_container_to_network', {
    networkId,
    containerId,
    ip,
  });
export const disconnectContainerFromNetwork = (
  networkId: string,
  containerId: string
) =>
  invoke<OkResponse>('disconnect_container_from_network', {
    networkId,
    containerId,
  });

// ─── STREAMING ────────────────────────────────────────────────────────────────
export interface LogChunk {
  containerId: string;
  containerName: string;
  stream: 'stdout' | 'stderr';
  message: string;
  timestamp: string;
}

export interface PullProgressChunk {
  image: string;
  tag: string;
  status: string;
  progress?: string;
  current?: number;
  total?: number;
  layerId?: string;
  done: boolean;
}

/**
 * Stream container logs to a callback.
 * Returns a cleanup function — call it to close the channel.
 *
 * @example
 * const stop = await streamContainerLogs('abc123', (chunk) => {
 *   console.log(chunk.message);
 * });
 * // later:
 * stop();
 */
export async function streamContainerLogs(
  id: string,
  onChunk: (chunk: LogChunk) => void,
  options?: { tail?: number; follow?: boolean }
): Promise<() => void> {
  const channel = new Channel<LogChunk>();
  channel.onmessage = onChunk;

  invoke('stream_container_logs', {
    id,
    tail: options?.tail,
    follow: options?.follow ?? true,
    channel,
  }).catch(console.error);

  return () => {
    // Closing the channel stops the Rust stream
    channel.onmessage = () => {};
  };
}

/**
 * Pull an image with streaming progress events.
 *
 * @example
 * const stop = await pullImageStream('nginx', 'alpine', (chunk) => {
 *   console.log(chunk.status, chunk.progress);
 * });
 */
export async function pullImageStream(
  image: string,
  tag: string,
  onProgress: (chunk: PullProgressChunk) => void
): Promise<() => void> {
  const channel = new Channel<PullProgressChunk>();
  channel.onmessage = onProgress;

  invoke('pull_image_stream', { image, tag, channel }).catch(console.error);

  return () => {
    channel.onmessage = () => {};
  };
}

// ─── EVENT LISTENERS ──────────────────────────────────────────────────────────

export interface DockerEvent {
  eventType: string;
  action: string;
  actorId: string;
  actorName: string;
  time: number;
  attributes: Record<string, string>;
}

/**
 * Listen for Docker daemon events (container start/stop/die etc).
 * Returns an unlisten function.
 *
 * @example
 * const unlisten = await onDockerEvent((event) => {
 *   if (event.action === 'die') refetchContainers();
 * });
 */
export async function onDockerEvent(
  handler: (event: DockerEvent) => void
): Promise<() => void> {
  const { listen } = await import('@tauri-apps/api/event');
  return listen<DockerEvent>('docker://event', (e) => handler(e.payload));
}

/**
 * Listen for bulk stats updates emitted by the stats collector every ~2s.
 */
export async function onDockerStats(
  handler: (stats: ContainerStats[]) => void
): Promise<() => void> {
  const { listen } = await import('@tauri-apps/api/event');
  return listen<ContainerStats[]>('docker://stats', (e) => handler(e.payload));
}

// ─── EXEC / TERMINAL ──────────────────────────────────────────────────────────

export interface ExecOutputChunk {
  sessionId: string;
  stream: 'stdout' | 'stderr';
  data: string;
}

/**
 * Start an interactive exec session inside a container.
 * Returns the session ID.
 * Output is streamed via the `exec://output` event — listen with
 * `onExecOutput`.
 */
export async function execSessionStart(
  containerId: string,
  shell: string = '/bin/sh'
): Promise<string> {
  return invoke<string>('exec_session_start', { containerId, shell });
}

/** Write data to the stdin of an exec session. */
export async function execSessionWrite(
  sessionId: string,
  data: string
): Promise<void> {
  return invoke('exec_session_write', { sessionId, data });
}

/** Resize the terminal inside the exec session. */
export async function execSessionResize(
  sessionId: string,
  cols: number,
  rows: number
): Promise<void> {
  return invoke('exec_session_resize', { sessionId, cols, rows });
}

/** Stop and clean up an exec session. */
export async function execSessionStop(
  sessionId: string
): Promise<void> {
  return invoke('exec_session_stop', { sessionId });
}

/** Listen for exec output events. Returns an unlisten function. */
export async function onExecOutput(
  handler: (chunk: ExecOutputChunk) => void
): Promise<() => void> {
  const { listen } = await import('@tauri-apps/api/event');
  return listen<ExecOutputChunk>('exec://output', (e) => handler(e.payload));
}

File: src-tauri/src/api/docker.rs
use crate::{
    api::utils::{map_err, CmdResult, CommandError, OkResponse, PruneResult},
    docker::{
        client::bytes_to_human,
        containers::ContainerOps,
        context::DockerContext,
        images::ImageOps,
        models::{
            ContainerInspect, ContainerStats, ContainerSummary, CreateNetworkOptions, DiskUsage,
            ImageSummary, NetworkSummary, RunContainerOptions, SystemInfo, VolumeSummary,
        },
        networks::NetworkOps,
        volumes::VolumeOps,
    },
    state::AppState,
};
use std::collections::HashMap;
use tauri::State;

// ═══════════════════════════════════════════════════════════════════════════════
// SYSTEM
// ═══════════════════════════════════════════════════════════════════════════════

/// Ping the Docker daemon. Returns `true` if reachable.
#[tauri::command]
pub async fn docker_ping(state: State<'_, AppState>) -> CmdResult<bool> {
    Ok(state.docker.ping().await)
}

/// Try to reconnect after the daemon was offline.
#[tauri::command]
pub async fn docker_reconnect(state: State<'_, AppState>) -> CmdResult<bool> {
    Ok(state.reconnect().await)
}

/// Fetch overall Docker system info.
#[tauri::command]
pub async fn docker_system_info(state: State<'_, AppState>) -> CmdResult<SystemInfo> {
    let docker = state.docker.get().await.map_err(CommandError::from)?;
    let (info, version) = tokio::try_join!(
        docker.info(),
        docker.version(),
    )
    .map_err(|e| CommandError::new(e.to_string()))?;

    Ok(SystemInfo {
        docker_version: info.server_version.clone().unwrap_or_default(),
        api_version: version.api_version.unwrap_or_default(),
        os: info.operating_system.clone().unwrap_or_default(),
        arch: info.architecture.unwrap_or_default(),
        total_memory: info.mem_total.unwrap_or(0) as u64,
        cpu_count: info.ncpu.unwrap_or(0) as u64,
        containers_running: info.containers_running.unwrap_or(0) as u64,
        containers_stopped: info.containers_stopped.unwrap_or(0) as u64,
        containers_paused: info.containers_paused.unwrap_or(0) as u64,
        images_count: info.images.unwrap_or(0) as u64,
        server_version: info.server_version.unwrap_or_default(),
        kernel_version: info.kernel_version.unwrap_or_default(),
        operating_system: info.operating_system.unwrap_or_default(),
        storage_driver: info.driver.unwrap_or_default(),
    })
}

/// Fetch disk usage summary.
#[tauri::command]
pub async fn docker_disk_usage(state: State<'_, AppState>) -> CmdResult<DiskUsage> {
    let docker = state.docker.get().await.map_err(CommandError::from)?;
    let df = docker
        .df()
        .await
        .map_err(|e| CommandError::new(e.to_string()))?;

    let containers_size = df
        .containers
        .unwrap_or_default()
        .iter()
        .filter_map(|c| c.size_root_fs) // This is Option<i64>, keep filter_map
        .map(|s| s.max(0) as u64)
        .sum();

    let images_size = df
        .images
        .unwrap_or_default()
        .iter()
        .map(|i| i.size) // i.size is i64, not Option, so use map instead of filter_map
        .map(|s| s.max(0) as u64) // Convert i64 to u64 safely
        .sum();

    let volumes_size = df
        .volumes
        .unwrap_or_default()
        .iter()
        .filter_map(|v| v.usage_data.as_ref()) // This is Option<UsageData>
        .map(|u| u.size) // u.size is i64, not Option
        .map(|s| s.max(0) as u64) // Convert i64 to u64 safely
        .sum();

    let build_cache_size = df
        .build_cache
        .unwrap_or_default()
        .iter()
        .filter_map(|b| b.size) // Check if this is Option<i64> or i64
        .map(|s| s.max(0) as u64)
        .sum();

    let total = containers_size + images_size + volumes_size + build_cache_size;

    Ok(DiskUsage {
        containers_size,
        images_size,
        volumes_size,
        build_cache_size,
        total_size: total,
    })
}

// ═══════════════════════════════════════════════════════════════════════════════
// CONTAINERS
// ═══════════════════════════════════════════════════════════════════════════════

#[tauri::command]
pub async fn list_containers(
    all: Option<bool>,
    state: State<'_, AppState>,
) -> CmdResult<Vec<ContainerSummary>> {
    let ops = ContainerOps::new(&state.docker);
    map_err(ops.list(all.unwrap_or(true)).await)
}

#[tauri::command]
pub async fn inspect_container(
    id: String,
    state: State<'_, AppState>,
) -> CmdResult<ContainerInspect> {
    let ops = ContainerOps::new(&state.docker);
    map_err(ops.inspect(&id).await)
}

#[tauri::command]
pub async fn start_container(id: String, state: State<'_, AppState>) -> CmdResult<OkResponse> {
    let ops = ContainerOps::new(&state.docker);
    map_err(ops.start(&id).await)?;
    Ok(OkResponse::with_message(format!(
        "Container {} started",
        id
    )))
}

#[tauri::command]
pub async fn stop_container(
    id: String,
    timeout: Option<i64>,
    state: State<'_, AppState>,
) -> CmdResult<OkResponse> {
    let ops = ContainerOps::new(&state.docker);
    map_err(ops.stop(&id, timeout).await)?;
    Ok(OkResponse::with_message(format!(
        "Container {} stopped",
        id
    )))
}

#[tauri::command]
pub async fn restart_container(
    id: String,
    timeout: i64,
    state: State<'_, AppState>,
) -> CmdResult<OkResponse> {
    let ops = ContainerOps::new(&state.docker);
    map_err(ops.restart(&id, timeout).await)?;
    Ok(OkResponse::with_message(format!(
        "Container {} restarted",
        id
    )))
}

#[tauri::command]
pub async fn pause_container(id: String, state: State<'_, AppState>) -> CmdResult<OkResponse> {
    let ops = ContainerOps::new(&state.docker);
    map_err(ops.pause(&id).await)?;
    Ok(OkResponse::with_message(format!("Container {} paused", id)))
}

#[tauri::command]
pub async fn unpause_container(id: String, state: State<'_, AppState>) -> CmdResult<OkResponse> {
    let ops = ContainerOps::new(&state.docker);
    map_err(ops.unpause(&id).await)?;
    Ok(OkResponse::with_message(format!(
        "Container {} unpaused",
        id
    )))
}

#[tauri::command]
pub async fn kill_container(
    id: String,
    signal: Option<String>,
    state: State<'_, AppState>,
) -> CmdResult<OkResponse> {
    let ops = ContainerOps::new(&state.docker);
    map_err(ops.kill(&id, &signal.unwrap_or_else(|| "SIGKILL".into())).await)?;
    Ok(OkResponse::with_message(format!(
        "Container {} killed",
        id
    )))
}

#[tauri::command]
pub async fn remove_container(
    id: String,
    force: Option<bool>,
    state: State<'_, AppState>,
) -> CmdResult<OkResponse> {
    let ops = ContainerOps::new(&state.docker);
    map_err(ops.remove(&id, force.unwrap_or(false)).await)?;
    Ok(OkResponse::with_message(format!(
        "Container {} removed",
        id
    )))
}

#[tauri::command]
pub async fn rename_container(
    id: String,
    new_name: String,
    state: State<'_, AppState>,
) -> CmdResult<OkResponse> {
    let ops = ContainerOps::new(&state.docker);
    map_err(ops.rename(&id, &new_name).await)?;
    Ok(OkResponse::ok())
}

#[tauri::command]
pub async fn run_container(
    options: RunContainerOptions,
    state: State<'_, AppState>,
) -> CmdResult<String> {
    let ops = ContainerOps::new(&state.docker);
    map_err(ops.run(options).await)
}

#[tauri::command]
pub async fn get_container_stats(
    id: String,
    state: State<'_, AppState>,
) -> CmdResult<ContainerStats> {
    let ops = ContainerOps::new(&state.docker);
    map_err(ops.stats(&id).await)
}

#[tauri::command]
pub async fn get_container_logs(
    id: String,
    tail: Option<u64>,
    since: Option<i64>,
    state: State<'_, AppState>,
) -> CmdResult<Vec<crate::docker::models::LogLine>> {
    let ops = ContainerOps::new(&state.docker);
    map_err(ops.logs(&id, tail, since).await)
}

// ═══════════════════════════════════════════════════════════════════════════════
// IMAGES
// ═══════════════════════════════════════════════════════════════════════════════

#[tauri::command]
pub async fn list_images(state: State<'_, AppState>) -> CmdResult<Vec<ImageSummary>> {
    let ops = ImageOps::new(&state.docker);
    map_err(ops.list().await)
}

#[tauri::command]
pub async fn inspect_image(id: String, state: State<'_, AppState>) -> CmdResult<ImageSummary> {
    let ops = ImageOps::new(&state.docker);
    map_err(ops.inspect(&id).await)
}

#[tauri::command]
pub async fn pull_image(
    image: String,
    tag: Option<String>,
    state: State<'_, AppState>,
) -> CmdResult<OkResponse> {
    let ops = ImageOps::new(&state.docker);
    let t = tag.as_deref().unwrap_or("latest");
    map_err(ops.pull(&image, t, |_| {}).await)?;
    Ok(OkResponse::with_message(format!("Pulled {}:{}", image, t)))
}

#[tauri::command]
pub async fn remove_image(
    id: String,
    force: Option<bool>,
    state: State<'_, AppState>,
) -> CmdResult<OkResponse> {
    let ops = ImageOps::new(&state.docker);
    map_err(ops.remove(&id, force.unwrap_or(false)).await)?;
    Ok(OkResponse::with_message(format!("Image {} removed", id)))
}

#[tauri::command]
pub async fn prune_images(state: State<'_, AppState>) -> CmdResult<PruneResult> {
    let ops = ImageOps::new(&state.docker);
    let (deleted, reclaimed) = map_err(ops.prune().await)?;
    Ok(PruneResult {
        deleted,
        reclaimed_bytes: reclaimed,
        reclaimed_human: bytes_to_human(reclaimed),
    })
}

// ═══════════════════════════════════════════════════════════════════════════════
// VOLUMES
// ═══════════════════════════════════════════════════════════════════════════════

#[tauri::command]
pub async fn list_volumes(state: State<'_, AppState>) -> CmdResult<Vec<VolumeSummary>> {
    let ops = VolumeOps::new(&state.docker);
    map_err(ops.list().await)
}

#[tauri::command]
pub async fn create_volume(
    name: String,
    driver: Option<String>,
    labels: Option<HashMap<String, String>>,
    driver_opts: Option<HashMap<String, String>>,
    state: State<'_, AppState>,
) -> CmdResult<VolumeSummary> {
    let ops = VolumeOps::new(&state.docker);
    map_err(
        ops.create(
            &name,
            driver.as_deref().unwrap_or("local"),
            labels.unwrap_or_default(),
            driver_opts.unwrap_or_default(),
        )
        .await,
    )
}

#[tauri::command]
pub async fn remove_volume(
    name: String,
    force: Option<bool>,
    state: State<'_, AppState>,
) -> CmdResult<OkResponse> {
    let ops = VolumeOps::new(&state.docker);
    map_err(ops.remove(&name, force.unwrap_or(false)).await)?;
    Ok(OkResponse::with_message(format!("Volume {} removed", name)))
}

#[tauri::command]
pub async fn prune_volumes(state: State<'_, AppState>) -> CmdResult<PruneResult> {
    let ops = VolumeOps::new(&state.docker);
    let (deleted, reclaimed) = map_err(ops.prune().await)?;
    Ok(PruneResult {
        deleted,
        reclaimed_bytes: reclaimed,
        reclaimed_human: bytes_to_human(reclaimed),
    })
}

// ═══════════════════════════════════════════════════════════════════════════════
// NETWORKS
// ═══════════════════════════════════════════════════════════════════════════════

#[tauri::command]
pub async fn list_networks(state: State<'_, AppState>) -> CmdResult<Vec<NetworkSummary>> {
    let ops = NetworkOps::new(&state.docker);
    map_err(ops.list().await)
}

#[tauri::command]
pub async fn create_network(
    options: CreateNetworkOptions,
    state: State<'_, AppState>,
) -> CmdResult<String> {
    let ops = NetworkOps::new(&state.docker);
    map_err(ops.create(options).await)
}

#[tauri::command]
pub async fn remove_network(id: String, state: State<'_, AppState>) -> CmdResult<OkResponse> {
    let ops = NetworkOps::new(&state.docker);
    map_err(ops.remove(&id).await)?;
    Ok(OkResponse::with_message(format!("Network {} removed", id)))
}

#[tauri::command]
pub async fn connect_container_to_network(
    network_id: String,
    container_id: String,
    ip: Option<String>,
    state: State<'_, AppState>,
) -> CmdResult<OkResponse> {
    let ops = NetworkOps::new(&state.docker);
    map_err(ops.connect(&network_id, &container_id, ip).await)?;
    Ok(OkResponse::ok())
}

#[tauri::command]
pub async fn disconnect_container_from_network(
    network_id: String,
    container_id: String,
    state: State<'_, AppState>,
) -> CmdResult<OkResponse> {
    let ops = NetworkOps::new(&state.docker);
    map_err(ops.disconnect(&network_id, &container_id).await)?;
    Ok(OkResponse::ok())
}

// ─── COMPOSE ─────────────────────────────────────────────────────────────────

#[tauri::command]
pub async fn compose_up(
    project_dir: String,
) -> CmdResult<OkResponse> {
    use std::process::Command;
    let output = Command::new("docker")
        .args(["compose", "-f", &format!("{}/docker-compose.yml", project_dir), "up", "-d"])
        .output()
        .map_err(|e| CommandError::from(e))?;
    if output.status.success() {
        Ok(OkResponse { ok: true, message: Some("Stack started".into()) })
    } else {
        let stderr = String::from_utf8_lossy(&output.stderr).to_string();
        Err(stderr.into())
    }
}

#[tauri::command]
pub async fn compose_down(
    project_dir: String,
) -> CmdResult<OkResponse> {
    use std::process::Command;
    let output = Command::new("docker")
        .args(["compose", "-f", &format!("{}/docker-compose.yml", project_dir), "down"])
        .output()
        .map_err(|e| CommandError::new(format!("Failed to run compose down: {}", e)))?;
    if output.status.success() {
        Ok(OkResponse { ok: true, message: Some("Stack stopped".into()) })
    } else {
        let stderr = String::from_utf8_lossy(&output.stderr).to_string();
        Err(stderr.into())
    }
}

// ─── BUILD ────────────────────────────────────────────────────────────────────

#[tauri::command]
pub async fn image_build(
    tag: String,
    dockerfile: String,
    context: String,
) -> CmdResult<OkResponse> {
    use std::process::Command;
    let output = Command::new("docker")
        .args(["build", "-t", &tag, "-f", &dockerfile, &context])
        .output()
        .map_err(|e| CommandError::new(format!("Failed to build image: {}", e)))?;
    if output.status.success() {
        Ok(OkResponse { ok: true, message: Some(format!("Built {}", tag)) })
    } else {
        let stderr = String::from_utf8_lossy(&output.stderr).to_string();
        Err(stderr.into())
    }
}

// ─── PUSH ─────────────────────────────────────────────────────────────────────

#[tauri::command]
pub async fn image_push(
    image: String,
    destination: String,
) -> CmdResult<OkResponse> {
    use std::process::Command;
    // Tag if different
    if image != destination {
        let tag_output = Command::new("docker")
            .args(["tag", &image, &destination])
            .output()
            .map_err(|e| CommandError::new(format!("Failed to tag image: {}", e)))?;
        if !tag_output.status.success() {
            let stderr = String::from_utf8_lossy(&tag_output.stderr).to_string();
            return Err(stderr.into());
        }
    }
    let output = Command::new("docker")
        .args(["push", &destination])
        .output()
        .map_err(|e| CommandError::new(format!("Failed to push image: {}", e)))?;
    if output.status.success() {
        Ok(OkResponse { ok: true, message: Some(format!("Pushed {}", destination)) })
    } else {
        let stderr = String::from_utf8_lossy(&output.stderr).to_string();
        Err(stderr.into())
    }
}

// ─── DOCKER CONTEXT ──────────────────────────────────────────────────────────

#[tauri::command]
pub async fn docker_context() -> CmdResult<DockerContext> {
    Ok(crate::docker::context::read_config().await)
}

// ─── REGISTRY CREDENTIALS (OS Keychain) ───────────────────────────────────────

#[tauri::command]
pub async fn registry_login(
    registry: String,
    username: String,
    password: String,
) -> CmdResult<OkResponse> {
    // Verify credentials by running docker login
    let verified = {
        use std::io::Write;
        use std::process::{Command, Stdio};
        let mut child = Command::new("docker")
            .args(["login", &registry, "-u", &username, "--password-stdin"])
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .spawn()
            .map_err(|e| CommandError::new(format!("Failed to spawn docker login: {}", e)))?;
        if let Some(stdin) = child.stdin.as_mut() {
            let _ = stdin.write_all(password.as_bytes());
        }
        let output = child
            .wait_with_output()
            .map_err(|e| CommandError::new(format!("docker login failed: {}", e)))?;
        if !output.status.success() {
            let stderr = String::from_utf8_lossy(&output.stderr).to_string();
            return Err(CommandError::new(stderr));
        }
    };

    // Store in OS keychain
    crate::docker::credentials::save(&registry, &username, &password)
        .map_err(CommandError::new)?;

    Ok(OkResponse {
        ok: true,
        message: Some(format!("Logged in to {}", registry)),
    })
}

#[tauri::command]
pub async fn registry_logout(registry: String) -> CmdResult<OkResponse> {
    crate::docker::credentials::delete(&registry).map_err(CommandError::new)?;
    Ok(OkResponse {
        ok: true,
        message: Some(format!("Logged out of {}", registry)),
    })
}

#[tauri::command]
pub async fn registry_list_credentials() -> CmdResult<Vec<crate::docker::credentials::StoredCredential>> {
    // Read stored registries from Docker config
    let ctx = crate::docker::context::read_config().await;
    let creds = ctx
        .auths
        .into_iter()
        .filter_map(|r| {
            crate::docker::credentials::get(&r)
                .ok()
                .or_else(|| {
                    Some(crate::docker::credentials::StoredCredential {
                        username: String::new(),
                        registry: r,
                    })
                })
        })
        .collect();
    Ok(creds)
}
